### ISSUE TITLE
Audit constant-time evaluation where appropriate

### ISSUE THREAD

**Issue by jasoncolburne** [WebOfTrust/cesride#116]
Some paths are more sensitive that others with respect to constant time evaluation, but it's often the application that determines when this is required. In particular, was want to avoid verifications that allow the adversary to glean knowledge about a particular operation due to differing server response times. [This](https://cryptopals.com/sets/4/challenges/31) is a cryptopals exercise with an example of such an attack.

Basically, exiting algorithms and loops early may allow an attacker to understand how to forge signatures or otherwise defeat a cryptosystem. Constant time comparisons and algorithms combat these attacks. Here is an example from the initial implementation of `Tholder` where we iterate through the keys in question and return early if any clause is not satisfied:

```rust
        let mut wio: usize = 0;
        for clause in clauses {
            let mut cw = Rational32::new(0, 1);
            for weight in clause {
                if sats[wio] {
                    cw += weight;
                }
                wio += 1;
            }
            if cw < *RATIONAL_ONE {
                return Ok(false);
            }
        }

        Ok(true)
```

which could instead be written:

```rust
        let mut result = Ok(true);
        let mut wio: usize = 0;
        for clause in clauses {
            let mut cw = Rational32::new(0, 1);
            for weight in clause {
                if sats[wio] {
                    cw += weight;
                }
                wio += 1;
            }
            if cw < *RATIONAL_ONE {
                result = Ok(false);
            }
        }

        result
```

The second approach returns in constant time. This situation would be exacerbated if we were doing a verification in this process, since adding time to the loop allows an adversary to more easily hone in on their target.

Important: ensure the compiler isn't optimizing away your work.

**Comment by jasoncolburne** [comment]
The example still isn't even really constant time - I suppose when thinking about it, for constant time functions every if should have a balanced else clause and we have two unbalanced ifs.

**Comment by SmithSamuelM** [comment]
@jasoncolburne Love seeing this level of introspection on the code for security.  Something to point out though. In general, constant timing on loops is a good idea for exactly the reason stated, so as not to disclose any information that might be usable to an attacker.  As you also know the main use case for timing attacks is to discover secrets, incrementally, bit by bit. When it comes to more brute force attacks like the whole signature, timing attacks are not so useful.  The tholder only runs after signatures are verified, which means an attacker has to first succeed in brute forcing the signature to pass verification before the tholder is ever called.  KERI accepts an event if it has one valid signature (and then places it into escrow). So a brute force attacker would attack the signature validation of a given event with a single signature. So KERI has to assume that signatures are not brute forceable. That the timing difference between accepting an event into escrow or rejecting the event because the signature was invalid does not materially weaken the security of the signature. 

So the question, is, any given signature by itself sufficiently unforgeable (i.e. computationally infeasible). Does timing of the acceptance of a packet or not based on its signature give an attacker enough information to forge a signature, like via a chosen message attack? If so then the code path for accepting with one signature is where timing matters. Not downstream of that. The security property of existential unforgeability means that knowleage of acceptance or not does not enable an attacker to forge a signature for a given key.   Moderan ECC signatures like ECDSA and EDDSA have EUA-CMA. One reason to use Ed25519 is that they have the property of strong unforgeablity which in addition to existential unforgeability means that an attacker can’t forge a message with the same signature but for a different key.  https://eprint.iacr.org/2020/823.pdf

I would be interested in your thoughts.

**Comment by SmithSamuelM** [comment]
KERI leverages EUA-CMA by never accepting an event that does not have a valid signature, so as not to be vulnerable to any other type of attack downstream in its processing. Validate signature first..  Because the only sigantures that are acceptable are from the keys in the key. list, EUA-CMA is sufficiet., But  KERI may be used for other applications where  SUA-CMA is ncessary.

**Comment by SmithSamuelM** [comment]
Just so anyone reading can follow.

When an AID is derived from the Inception Event as a digest of that event, then CMA on the inception event will fail because the AID will be different for each chosen message.  This includes the initial set of public keys. So EUA-CMA is sufficient for. a given AID, because forging a signature for a different key pair to the same inception event does not help with any subsequent events so the attacker can’t gain control of the AID.  But SUA-CMA on the inception event means that an attacker who wants to attack a multi-sig group can’t get the inception accepted inadvertently. So SUA-CMA is better. Becasue EUA-CMA requires on OOB authentication of the intent of the key holders to prevent a valid signature using the attacker  key pair being accepted as valid for a member’s valid key pair.

The pre-rotation mechanism means that the next set of valid keys is committed too using a hash of each as part of the inception event. So a CMA attack on the inception event to attack the next keys would change the AID of the inception event.  Without knowledge of the actual next public keys an attacker can’t construct a CMA on a subsequent rotation event until its too late. But once again, given a valid rotation event with exposed signing keys, EUA-CMA, can’t gain control of the AID because any subsequent events would fail. And the only attack would be for a rotation event that would otherwise be rejected by the groupb. But SUA-CMA is better because an attacker can’t create a valid signature even for a different key pair so the OOB step is not needed.  

So one reason to always sign every packet is so that all downstream processing is protected by at least EUA-CMA. This enables KERI to make performance tradeoffs in the down-stream processing that would not be secure otherwise.

**Comment by jasoncolburne** [comment]
> When it comes to more brute force attacks like the whole signature, timing attacks are not so useful.

You are right, I was thinking of some attack where an adversary had all the keys but did not know the order in which they were required, which seems a bit silly but still conceivable if an application is improperly implemented. Should we assume all applications are protected?

You are correct though, in this case where the signature is not being verified it is likely to be so fast that in a remote scenario even that attack would be infeasible.

**Comment by jasoncolburne** [comment]
I am thinking for non-establishment events, doesn't thresholded signing work there too?

**Comment by jasoncolburne** [comment]
Would it be fair to say then, @SmithSamuelM that `cesride` is fairly immune to these type of attacks in general given that the underlying crypto implements constant-time solutions?

**Comment by jasoncolburne** [comment]
>I am thinking for non-establishment events, doesn't thresholded signing work there too? Low volume interactions that go unnoticed - duplicity detection may happen but too late? Again I don't think this is a reasonable attack, and maybe I've missed something above that makes it impossible.

I guess the attacker would always know the index of each key making that all moot? Meaning that any adversary in that position has already won.

My question about the library's general immunity remains though.. Are we really only vulnerable at present when doing verifications? And since we are relying on another library to do the comparisons we just need to ensure the dependencies are implemented with constant time in mind?

**Comment by jasoncolburne** [comment]
Oh I missed what you said in your original message, Tholder runs _after_ verification. I now remember this from a meeting. Thanks. So my next question is, during verification, do we have a similar loop in our architecture that exits early? I am not sure exactly where to look in KERIpy for that code, but I can do some searching.

**Comment by SmithSamuelM** [comment]
The signature verification loop has two modes. When the message itself has attached enough signatures to meet the threshold and when it does not.  In the first case, all the attached signatures are verified always because even though only one verified signature is enough to store in escrow, we want to store all verified signature even if they exceed the threshold. Indeed any message that is accepted due to a threshold will still verify any new signatures and store them with the message. This is too support duplicity accountability by any of the signers. So all attached signatures are always verified. It does not exit the verification loop early.  In the second case the escrow processing which happens asynchronosly to message arrival (which means the attacker has no knowledge) just examines all of the so far escrowed verified  signatures using the tholder to see if they meet the threshold.  

> My question about the library's general immunity remains though.. Are we really only vulnerable at present when doing verifications? And since we are relying on another library to do the verification we just need to ensure the dependencies are implemented with constant time in mind?

So as long as the actual signature verification library (libsodium for example) is not timing sentitive to attack, the overall verification is protected by EUA-CMA regardless of the timing sensitivity of the thresholding mechanism. But as mentioned above we always verify all attached signatures so there is no timing sensitivity to the verification loop.

**Comment by SmithSamuelM** [comment]
@jasoncolburne  

> Would it be fair to say then, @SmithSamuelM that cesride is fairly immune to these type of attacks in general if we assume that the underlying crypto implements constant-time solutions?

Yes, the sign everything approach protects any down stream processing as long as the actual signature verification library calls are timing insensitive. Anyone using a digital signing library that has timing sensitive signature verification should be exiled.

**Comment by jasoncolburne** [comment]
Okay awesome, I'll rework this ticket to be more about auditing our dependencies. Thanks @SmithSamuelM !
